presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -212,7 +212,11 @@ private SqlStageExecution(
     // this is a separate method to ensure that the `this` reference is not leaked during construction
     private void initialize()
     {
-        stateMachine.addStateChangeListener(newState -> checkAllTaskFinal());
+        stateMachine.addStateChangeListener(newState -> {
+            if (newState.isDone()) {
```

Comment:
Do you need this newState.isDone check given that stateMachine.getState().isDone() already exists in checkAllTasksFinal ?

Reply:
this check is outside the synchronized block, so we're again the reducing the number of times checkAllTasksFinal needs to be called and the lock acquired. At this point we know that the trigger is moving to a new state. We don't need to worry about whether there is another concurrent update to the state that moves it to a done state, as that will trigger an additional listener call where this check will pass and we will call checkAllTasksFinal.